src/libraries/Microsoft.PowerFx.Interpreter/RecalcEngine.cs

```diff
@@ -159,9 +166,14 @@ public async Task<FormulaValue> EvalAsync(string expressionText, CancellationTok
 
             var check = Check(expressionText, (RecordType)parameters.IRContext.ResultType, options);
             check.ThrowOnErrors();
-
-            var newValue = await check.Expression.EvalAsync(parameters, cancel);
-            return newValue;
+            try
+            {
+                return await check.Expression.EvalAsync(parameters, cancel);
+            }
+            catch (MaxCallDepthException)
+            {
+                return CommonErrors.MaxCallDepth(new IRContext(new Syntax.Span(0, expressionText.Length - 1), FormulaType.BindingError));

```

Comment:
> return CommonErrors.MaxCallDepth(new IRContext(new Syntax.Span(0, expressionText.Length - 1), FormulaType.BindingError)); [](http://example.com/codeflow?start=15&length=122)

move as helper on MaxCallDepthException

Also, if we're passing in a source context, we should have a text that verifies the source span is what we expect. .

Reply:
Unnecessary now that we have access to IRContext.